include/ensmallen_bits/pso/init_policies/constrained_init.hpp

```diff
@@ -0,0 +1,126 @@
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef ENSMALLEN_PSO_INIT_POLICIES_CONSTRAINED_INIT_HPP
+#define ENSMALLEN_PSO_INIT_POLICIES_CONSTRAINED_INIT_HPP
+
+namespace ens {
+
+/**
+ * The default initialization policy used by the PSO optimizer. It initializes
+ * particle positions uniformly in [-1, 1], the velocities in [0, 1] personal
+ * bests of the particles to the initial positions, and all fitness values to
+ * std::numeric_limits<double>::max().
+ */
+class ConstrainedInit
+{
+ public:
+  /**
+   * Constructor for the DefaultInit policy. The policy initializes particle
+   * posiitons in the range [lowerBound, upperBound]. Defaults to [-1, 1].
+   *
+   * @param lowerBound Lower bound of the position initialization range.
+   * @param upperBound Upper bound of the position initialization range.
+   */
+  ConstrainedInit(const double lowerBound = -1.0,
+              const double upperBound = 1.0):
+      lowerBound(lowerBound),
+      upperBound(upperBound)
+  {
+    /* Nothing to do */
+  }
+
+  /**
+   * The InitializeParticles method of the init policy. Any class that is used
+   * in place of this default must implement this method which is used by the
+   * optimizer.
+   *
+   * @param iterate Coordinates of the initial point for training.
+   * @param numParticles The number of particles in the swarm.
+   * @param particlePositions Current positions of particles.
+   * @param particleVelocities Current velocities of particles.
+   * @param particleFitnesses Current fitness values of particles.
+   * @param particleBestPositions Best positions attained by each particle.
+   * @param particleBestFitnesses Best fitness values attained by each particle.
+   */
+  template<typename FunctionType>
+  void InitializeParticles(FunctionType& function,
+			   const arma::mat& iterate,
+                           const size_t numParticles,
+                           arma::cube& particlePositions,
+                           arma::cube& particleVelocities,
+                           arma::mat& particleFitnesses,
+                           arma::cube& particleBestPositions,
+                           arma::mat& particleBestFitnesses)
+  {
+    // Randomly initialize the particle positions.
+    particlePositions.randu(iterate.n_rows, iterate.n_cols, numParticles);
+    // User provided weights replacement performed here.
+    arma::mat positionVector;
+    double positionFitness;
+    double inf = std::numeric_limits<double>::max();
+    for (size_t i = 0; i < numParticles; i++)
+    {
+      bool positionFeasible = 0; //Flag for tracking if generated position vector is feasible or not
+      while (positionFeasible != 0)
+      {
+        //Generate random position vector
+        positionVector.randu(iterate.n_rows,iterate.n_cols);
```

Comment:
We should generate the position around the starting point (``iterate``), so ``positionVector = arma::randu(iterate.n_rows,iterate.n_cols) + iterate``.

Reply:
I was thinking to incorporate the `min` and `max` values of the initial ranges along each axis into the particle initialization procedure. That will also be useful for initializing particles for constrained optimization.

Reply:
That works as well.